File: leetcode/oop/singleton.py
class Singleton:
    _instance = None

    def __new__(cls1, *args, **kwargs):
        if not cls1._instance:
            cls1._instance = super(Singleton, cls1).__new__(cls1, *args, **kwargs)
        return cls1._instance
    
    def __init__(self1):
        self1.value = 42

singleton1 = Singleton()
singleton2 = Singleton()

print(singleton1 is singleton2)  # True
print(singleton1.value)  # 42

File: leetcode/oop/Singleton.cpp
#include <iostream>

class Singleton {
    private:
    static Singleton* _instance;

    // Private constructor to prevent instantiation
    Singleton() {
        value = 42;
    }

    public:
    static Singleton* getInstance() {
        if (_instance == nullptr) {
            _instance = new Singleton();
        }
        return _instance;
    }

    int getValue() const {
        return value;
    }

    private:
    int value;
};

Singleton* Singleton::_instance = nullptr;

int main() {
    Singleton* singleton1 = Singleton::getInstance();
    Singleton* singleton2 = Singleton::getInstance();

    std::cout << (singleton1 == singleton2) << std::endl;  // true
    std::cout << singleton1->getValue() << std::endl;      // 42

    return 0;
}
